### DISCUSSION TITLE
React Native, wrapping the Controller loses focus on change...

### DISCUSSION THREAD

**Discussion by mattwoberts** [react-hook-form/react-hook-form#8168] [Q&A] [answered]
H!

Using this lovely lib to validate forms in our app... It works like a charm in react native. However I have a screen that contains a lot of text-based fields, so I tried to create a helper component to wrap them like this:

```javascript
const WrappedTextField = ({fieldName, placeHolder}) => (
    <View style={styles.item}>
    <Controller
      control={control}
      rules={{ required: false, maxLength: 200 }}
      render={({ field: { onChange, value } }) => (
        <TextInput
          style={styles.input}
          onChangeText={(val) => { dispatch({ type: 'DATA_CHANGED' }); onChange(val); }}
          value={value}
          placeholder={placeHolder}></TextInput>
      )}
      name={fieldName}
    />
  </View>
  )
```

And calling it for example like this:

`                <WrappedTextField fieldName='jobTitle' placeHolder='Job Title'/>`

It works - but as soon as you type into the text input, it loses focus with each key press. 

I'm not sure if this is an issue with the lib, or more likely an issue with me doing something wrong in react....

**Comment by bluebill1049** [comment, marked as answer]
Are you putting that component inside the component? This will lead the component gets mount and unmount each render.

**Reply by mattwoberts** [reply]
Argh yes I am - I'm fairly new to React, these things trip me up sometimes... I guess I can wrap with a `useCallback` for the wrapped component to avoid it from happening.. Thanks for that!

**Reply by bluebill1049** [reply]
I would move it out.

**Reply by Moshyfawn** [reply]
Yea, there's usually no need to define your component inside another component as you can simply pass the required props to it via.. well, the props API, per say. Think pure functions with no side-effects.

In fact, wrapping your components in `useCallback` is incorrect and would tank your app performance as it grows.

You can memoize your component with `React.memo`, but you need to do so in a very small percentage of cases ;)

**Reply by mattwoberts** [reply]
Thanks all for the advice - I've moved it up out of the containing component and all works well  :)
